### ISSUE TITLE
CMake is slow and generates the test files every time it is run

### ISSUE THREAD

**Issue by fruffy** [p4lang/p4c#4542]
Our CMake setup is rather slow. The slowness comes from the test files we generate unconditionally with the `p4c_add_tests` every time time we run `cmake`. It would be nice to find a way to cache the generation or only regenerate files when necessary.

**Comment by Sanket-0510** [comment]
there is a command in cmake `configure_file()` it kind of helps in caching a file [configure_file()](https://cmake.org/cmake/help/latest/command/configure_file.html)  I see [p4c_add_tests macro](https://github.com/p4lang/p4c/blob/e0c4b8a77302422366e7a34d379dfe2531a5a7b4/cmake/P4CUtils.cmake#L261C1-L267C24)   will this a good to to cache the generation of these test files - [files](https://github.com/p4lang/p4c/blob/e0c4b8a77302422366e7a34d379dfe2531a5a7b4/backends/ebpf/CMakeLists.txt#L242C1-L243C104)

**Comment by fruffy** [comment]
Yes, it could work. The problem is that the often generate test files in incremental fashion (by appending to the file). We would need to change that and emit the string in its entirety.

**Comment by RoshanKumar9421** [comment]
Hi, I’m preparing for Google Summer of Code and have built p4c locally.

I’d like to work on this issue. My plan is to identify the test generation
CMake targets and convert them to use proper OUTPUT/DEPENDS so tests are
only regenerated when inputs change.

Please let me know if this approach sounds good or if there’s a preferred direction.
Thanks!

**Comment by fruffy** [comment]
That sounds reasonable. Start small with some specific tests, for example the eBPF tests: https://github.com/p4lang/p4c/blob/main/backends/ebpf/CMakeLists.txt#L245

In general, you will have to improve the functionality of `p4c_add_tests`. You could introduce a `p4c_add_tests_2` with the same interface but the improve behavior for now.

**Comment by Harsimran-Dalal** [comment]
Hi,  
I’m interested in working on this issue.  
Could you confirm the expected behavior or point me to relevant parts of the codebase? Thanks!

**Comment by fruffy** [comment]
> Hi, I’m interested in working on this issue. Could you confirm the expected behavior or point me to relevant parts of the codebase? Thanks!

The previous comments should contain all the context you need. You also expressed interest in #4534, I would decide which issue to take on.

**Comment by nishant11-tchy** [comment]
Hi, I’d like to work on this issue. 

I’ll look into how test files are generated in the p4_add_tests macro and try to avoid unnecessary regeneration.
